### DISCUSSION TITLE
Layer repeats via a config file.

### DISCUSSION THREAD

**Discussion by dnhkng** [turboderp/exllamav2#270] [General]
I really liked the discussion on Reddit [localllama](https://www.reddit.com/r/LocalLLaMA/comments/190zsk1/comment/kgu7ffj/) about the potential of easily creating Frankenstein models using exllama.

I think that would open up new areas of research without huge compute and memory requirements. 

Could we start a discussion on 1) minimal Proof of Concept code modifications to test the ideas, and if the results are positive, 2) how we could make the inference efficient. 

On the first topic, where in codebase would I start?

**Comment by bjj** [comment]
@epolewsky did this in https://github.com/epolewski/EricLLM

**Reply by dnhkng** [reply]
Since I posted, it's been added to a few forks of Oobabooga (mine included).

**Comment by turboderp** [comment]
I actually had my own version of this in mind before the recent PR. Not sure which direction to take this right now. I'm a little hesitant because the idea of *easily* creating Frankenstein models puts me off a little bit. I mean, I wouldn't be too surprised if HF starts banning merged models soon because they don't actually have an unlimited budget for hosting all the files.

I've yet to hear a compelling reason for why Frankenstein models should work in the first place, and I've yet to see an objective (or blind) test showing that they do. But they're definitely crowding the space, so... idk.

**Reply by dnhkng** [reply]
I have objective results, just writing up a preprint.

Using various LLMs as judges (had to develop a good LLM Evaluation system in the process of doing this), there does appear to be benefits.

Happy to chat about the results privately.

I would like to have tried it on larger models, but the biggest I can test is on Yi34B models due to passing the data between GPUs on my dual 4090 system.  I assume addressing that issue is what you mean with 'your own version'?

**Reply by silphendio** [reply]
What do you think about incorporating LoRAs?

I posted my own version of this on reddit, and  [someone asked ](https://www.reddit.com/r/LocalLLaMA/comments/1aqrd7t/comment/kqfpb32/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) about the "possibility of laying a Lora trained on the static merge (say a 120B) on top of a dynamic 70B merge".

I think this could be a good next step.

But after taking a closer look at the source, hacking together a solution would be a bit more involved that just copying layers and changing `layer_idx`.

**Comment by dnhkng** [comment]
@turboderp 

This method fails on the new dynamic generator, and something has broken in the legacy system too.
I am (very slowly) in the process of writing a paper on the topic, and I would like to repeat the process with some newer models.

Do you have any suggestions on how I could do layer repeats using the current code base?

I would like to share some private data with you on the topic of the paper, please contact me at either:
https://www.linkedin.com/in/dnhkng/
or dnhkng @ gmail dot com

> I've yet to hear a compelling reason for why Frankenstein models should work in the first place, and I've yet to see an objective (or blind) test showing that they do. But they're definitely crowding the space, so... idk.

I can provide some compelling data.
